More configuration options for storage daemon

erik-frontify opened the pull request:
Allow defining storage daemon settings with two new hashes.

storage_device_hash
This allows to configure storage devices on the storage daemon using a
hiera or puppet hash.
Should be backward compatible as it uses the former default as default
hash.

storage_hash (which is an array of hashes, actually)
As there are some new resources in bacula 9.4 which aren't covered by
this module I created this array of hashes that allows to configure any
resource which might be needed. It's an array as a hash would not allow
to define one resource type twice as hash keys are merged if not unique.

blackknight36 commented:
Thanks.  I'll have to test this on our bacula director to make sure it doesn't cause any problems but I can do that later today.

Also, can you rebase and squash this into one commit?

erik-frontify commented:
OK, squashed them into one commit, now. I already squashed my commits before into the two main commits so you can see both changes ;-)
I was thinking about rewriting the module to better match the current defaults provided by bacula. But that would for sure break backwards compatibility (like removing most of the currently created job definitions).  And I'm not sure if I'll be having enough time to do so...

blackknight36 commented:
> I was thinking about rewriting the module to better match the current defaults provided by bacula. But that would for sure break backwards compatibility (like removing most of the currently created job definitions).

Changing the defaults would most likely require a major version bump however I'm not opposed to doing so as long as we still have the ability to override the values in hiera.  Time is also a concern as always.

blackknight36 commented on manifests/init.pp:
```diff
@@ -287,6 +301,8 @@
   Boolean $manage_logwatch               = true,
   String $operator_command               = "/usr/sbin/bsmtp -h localhost -f bacula@${::fqdn} -s \\\"Bacula Intervention Required (for %c)\\\" %r",
   String $plugin_dir                     = '/usr/lib64/bacula',
+  Optional[Array] $storage_hash           = undef,
```
This should be set to Optional[Hash] $storage_hash to avoid puppet failures.  Running this code on our director results in an error as follows.

```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Class[Bacula::Storage]: parameter 'storage_hash' expects an Array value, got Hash (file: /etc/puppetlabs/code/environments/bacula_test/forge_modules/bacula/manifests/init.pp, line: 391, column: 5) on node bacula.example.com
```

erik-frontify replied:
Good catch. I forgot to adjust the default value to an empty array.
I used a hash before until realising that that wont work.

blackknight36 replied:
No problem.  Once thing I've noticed is that the default value for Block Checksum now evaluates as "yes" even when I set the parameter to false in hiera.  I believe this might be because there is no "false" value set in the selector.

```
   'Block Checksum'   => $block_checksum ? {
      default => yes,
      no      => no,
    },
```

This should most likely be set as follows.

```
   'Block Checksum'   => $block_checksum ? {
      default => 'yes',
      false   => 'no',
    },
```

Puppet does not treat yes/no as "truthy" values, they are simply strings.

blackknight36 replied:
This module previously treated yes/no values for bacula settings as strings however I converted it to use booleans for each parameter instead.  The puppet templates then evaluate these booleans to fill in the proper yes/no values.

blackknight36 replied:
I've fixed this with commit 9c0a1e2

blackknight36 commented:
I've verified that the changes work on our bacula server.   I'll go ahead and merge this.
